crates/credentials/src/vc.rs

```diff
@@ -127,25 +195,163 @@ impl VerifiableCredential {
         Ok(jwt)
     }
 
-    pub async fn verify(jwt: &str) -> Result<Self, CredentialError> {
+    pub async fn verify(jwt: &str) -> Result<Arc<VerifiableCredential>, CredentialError> {
         let jwt_decoded = Jwt::verify::<VcJwtClaims>(jwt).await?;
+        let mut vc = jwt_decoded.claims.vc;
+        let registered_claims = jwt_decoded.claims.registered_claims;
 
-        // TODO Implement semantic VC verification rules https://github.com/TBD54566975/web5-rs/issues/151
+        // registered claims checks
+        if registered_claims.issuer.is_none() {
+            return Err(CredentialError::MissingIss);
+        }
+        if registered_claims.subject.is_none() {
+            return Err(CredentialError::MissingSub);
+        }
+        if registered_claims.not_before.is_none() {
+            return Err(CredentialError::MissingNbf);
+        }
+        if registered_claims.jti.is_none() {
+            return Err(CredentialError::MissingJti);
+        }
 
-        Ok(jwt_decoded.claims.vc)
+        // TODO: Change this to None after this issue is resolved - https://github.com/TBD54566975/web5-rs/issues/202
+        // check claims match expected values
+        if vc.id == String::default() {
+            vc.id = registered_claims.jti.clone().unwrap();
+        } else if vc.id != registered_claims.jti.unwrap() {
+            return Err(CredentialError::IdMismatch);
+        }
+
+        if vc.credential_subject.id == String::default() {
+            vc.credential_subject.id = registered_claims.subject.clone().unwrap();
+        } else if vc.credential_subject.id != registered_claims.subject.unwrap() {
+            return Err(CredentialError::SubjectMismatch);
+        }
+
+        if vc.issuance_date == i64::default() {
+            vc.issuance_date = registered_claims.not_before.unwrap();
+        } else if vc.issuance_date != registered_claims.not_before.unwrap() {
+            return Err(CredentialError::IssuanceDateMismatch);
+        }
+
+        if vc.expiration_date.unwrap() == i64::default() {
+            vc.expiration_date = Some(registered_claims.expiration.unwrap())
+        } else if vc.expiration_date != Some(registered_claims.expiration.unwrap()) {
+            return Err(CredentialError::ExpirationDateMismatch);
+        }
+
+        // issuer check
+        let iss = registered_claims.issuer.as_ref().unwrap();
+        match &mut vc.issuer {
+            Issuer::String(id) => {
+                if id.is_empty() {
+                    id.clone_from(iss);
+                } else if id != iss {
+                    return Err(CredentialError::IssuerMismatch);
+                }
+            }
+            Issuer::Object(named) => {
+                if &named.id != iss {
+                    return Err(CredentialError::IssuerMismatch);
+                }
+            }
+        }
```

Comment:
Nothing wrong with yours but I find the `&mut` to be a bit esoteric to follow, this may be simpler (once again you could remove the `if registered_claims.issuer.is_none()` condition above b/c that occurs here inline)

```suggestion
        let iss = registered_claims
            .issuer
            .ok_or(CredentialError::MissingIss)?;
        let vc_issuer = vc.issuer.to_string();
        if vc_issuer.is_empty() {
            vc.issuer = Issuer::String(iss.clone());
        } else if iss != vc_issuer {
            return Err(CredentialError::IssuerMismatch);
        }
```